### ISSUE TITLE
Compilation warnings and failure after erlang update

### ISSUE THREAD

**Issue by lukaszsamson** [elixir-lang/elixir#7124]
After updating erlang from 20.1 to 20.2 compile is noisy:
```
$ mix phx.server
==> ex2ms
Compiling 1 file (.ex)
warning: redefining module Ex2ms (current version loaded from /Users/app/_build/dev/lib/ex2ms/ebin/Elixir.Ex2ms.beam)
  lib/ex2ms.ex:1

Generated ex2ms app
==> base64url (compile)
==> file_system
Compiling 6 files (.ex)
warning: redefining module FileSystem (current version loaded from /Users/app/_build/dev/lib/file_system/ebin/Elixir.FileSystem.beam)
  lib/file_system.ex:1

warning: redefining module FileSystem.Worker (current version loaded from /Users/app/_build/dev/lib/file_system/ebin/Elixir.FileSystem.Worker.beam)
  lib/file_system/worker.ex:1

warning: redefining module FileSystem.Backend (current version loaded from /Users/app/_build/dev/lib/file_system/ebin/Elixir.FileSystem.Backend.beam)
  lib/file_system/backend.ex:3
```
This goes on for every dependancy and module.

Finally, the build fails or completes but fails to start. In this case it failed with
```
== Compilation error in file lib/gettext/extractor.ex ==
** (KeyError) key :extracted_comments not found in: %Gettext.PO.PluralTranslation{comments: [], flags: {{:., [line: 131], [MapSet, :new]}, [line: 131], [["elixir-format"]]}, msgid: [{:msgid, [line: 128], nil}], msgid_plural: [{:msgid_plural, [line: 129], nil}], msgstr: {:%{}, [line: 130], [{0, [""]}, {1, [""]}]}, po_source_line: nil, references: [{{{:., [line: 132], [Path, :relative_to_cwd]}, [line: 132], [{:file, [line: 132], nil}]}, {:line, [line: 132], nil}}]}
    (stdlib) :maps.update(:extracted_comments, {:extracted_comments, [line: 133], nil}, %Gettext.PO.PluralTranslation{comments: [], flags: {{:., [line: 131], [MapSet, :new]}, [line: 131], [["elixir-format"]]}, msgid: [{:msgid, [line: 128], nil}], msgid_plural: [{:msgid_plural, [line: 129], nil}], msgstr: {:%{}, [line: 130], [{0, [""]}, {1, [""]}]}, po_source_line: nil, references: [{{{:., [line: 132], [Path, :relative_to_cwd]}, [line: 132], [{:file, [line: 132], nil}]}, {:line, [line: 132], nil}}]})
    lib/gettext/po/plural_translation.ex:40: anonymous fn/2 in Gettext.PO.PluralTranslation.__struct__/1
    (elixir) lib/enum.ex:1896: Enum."-reduce/3-lists^foldl/2-0-"/3
could not compile dependency :gettext, "mix compile" failed. You can recompile this dependency with "mix deps.compile gettext", update it with "mix deps.update gettext" or clean it with "mix deps.clean gettext"
```
The suggested solution `mix deps.compile gettext` obviously does not fix the problem
`mix clean` does not help either
`mix deps.clean --all` fixes the problem

### Environment

* Elixir & Erlang versions: erlang 20.2, elixir 1.6-dev
* Operating system: macOS 10.13.2

### Current behavior

as described above

### Expected behavior

mix should detect that erlang version changed and clean deps properly

**Comment by josevalim** [comment]
This is expected for Elixir v1.6-dev. We keep a manifest of precompiled files but we don't always bump the manifest version in development versions. By the time v1.6 is out, then the manifest is bumped between v1.5 and v1.6, and then everything should be cleaned properly.

TL;DR - it may happen when running on Elixir master, it should not happen between stable versions.
